### ISSUE TITLE
Option "Process during tracking request" is ignored

### ISSUE THREAD

**Issue by Programie** [matomo-org/plugin-QueuedTracking#113]
I tried to enable this plugin in our Matomo setup which now seems to be work fine except for one issue: The option "Process during tracking request" seems to be ignored and processing is always done during the tracking request.

The cronjob calling "console queuedtracking:process" always responses with "0 requests in 0.01 seconds" and the number of items in the "trackingQueueV1" Redis key always change.

I would expect, that disabling the cronjob should result in an increase of the queue which does not happen for me.

**Comment by tsteur** [comment]
From what I can see everything works as expected. If you don't want requests to be processed during tracking maybe make sure again it is disabled. And make sure that queued tracking in general is enabled and that requests are written into the queue. 

Maybe check what `./console queuedtracking:monitor` and `./console queuedtracking:test` says. Eg the monitor will tell you in the beginning if process during tracking is enabled or whether the command to process needs to be executed.

**Comment by Programie** [comment]
`./console queuedtracking:monitor` shows that it is processing the queue as it is processed during the tracking request. I guess the number of items in the queue shouldn't decrease all the time if the cronjob is only executed once per minute.

The output of `./console queuedtracking:test` also looks fine (no errors or warnings).

**Comment by tsteur** [comment]
You might need to increase the number of how many requests should be processed in parallel. Not quite sure what your goal is and what you expect to see.

**Comment by tsteur** [comment]
If you increase the number of requests to process, you should see a higher number in the queue depending on amount of traffic.

**Comment by Programie** [comment]
My goal: I want to use this plugin to prevent any database access done by the tracking requests to allow easier maintenance (e.g. disable the cronjob to allow database downtime, migration, etc. while keeping the tracking data in the Redis as some kind of buffer).

If I understand it correctly, disabling the option "Process during tracking request" should result in the tracking data not being written to the database and kept in the Redis, which is not the case. The data is always written to the database even if the option is disabled and the cronjob is not running.

So this looks like a bug to me.

**Comment by tsteur** [comment]
Not sure which data you mean? You mean the tracking data? If you have configured Redis it should only write the data into redis and not process the queue immediately if the "processing of the queue during tracking is disabled". When it is disabled, it will only process through CLI. Simply disable all cronjobs for a while and look into the monitor to see what happens. For us it's working correctly as expected.

**Comment by Programie** [comment]
That's exactly what I did. I've configured the plugin to not process the queue immediately but still the queue is processed without the console command being invoked.

So, my cronjob which would usually start `console queuedtracking:process` every minute is currently DISABLED. The option "Process during tracking request" is DISABLED. The tracking data is still processed (i.e. I can see up-to-date visits in Matomo and the queue in the Redis is increasing and decreasing all the time).

**Comment by tsteur** [comment]
In https://github.com/matomo-org/plugin-QueuedTracking/blob/3.3.5/Tracker/Handler.php#L80 can you add a `return;` right at the beginning of the method before `Common::printDebug` and see if it changes something? 

I cannot reproduce this unfortunately and looking at the code I would not be able to explain what is happening there. Maybe also double check with `ps` or something that no `process` command is running in the background etc and if you use multiple servers that the process is stopped on all servers.

**Comment by Programie** [comment]
Adding the `return;` to the `processQueue` method prevents the queue from being processed in the tracking requests while still allowing queue processing from the command line.

So I guess, the method should actually check whether the option has been activated or not and do nothing if it is disabled.

**Comment by tsteur** [comment]
This is already checked here: https://github.com/matomo-org/plugin-QueuedTracking/blob/3.3.5/QueuedTracking.php#L62-L63

For some reason your setting seems to evaluate to true here. It's a boolean field so for some reason it thinks it's enabled for you. Can you check there is also no config setting for `processDuringTrackingRequest` in your config/config.ini.php file etc?

**Comment by Programie** [comment]
I did some more debugging but now it looks like the queue is no longer processed if the cronjob is not running. So everything works as expected and I still don't know why it didn't worked before.

I also did a full reset of all files to the ones from the Matomo and QueuedTracking distribution files but it still works as expected.

TL;DR: I don't know what went wrong, but now it's working as expected.

**Comment by tsteur** [comment]
Thanks for letting us know 👍

**Comment by pardvm** [comment]
Hi @tsteur,
I think I've found a situation very closed to the one exposed by @Programie and I guess that it has to do with the fact that `processDuringTrackingRequest` setting is disabled from the UI settings panel that, **in my case**, is running in a different computer than the one where the tracking requests are being served.

If both, the UI Settings and the tracking requests are served from the same computer, then, I cannot reproduce de effect... Does it has any sense?

**Comment by Programie** [comment]
In my case, Matomo is deployed to a cluster of 6 servers running behind a load balancer (balancing in round robin style). I've made sure that the config.ini.php has the same content on all servers.

**Comment by pardvm** [comment]
Same here: a load balancer in front of several matomo servers. One of those servers handles all UI requests and others handle only tracking requests
In order to disable `processDuringTrackingRequest`, I need to specifically navigate to the UI running in the node that is handling tracking requests.

**Comment by tsteur** [comment]
We don't have any problems with such a setup. If you configure this in the config file, you just indeed need to make sure to have the same config file on all servers. I can't think of a reason this should cause any problem / difference. Make sure the config file is the same on all servers and they run they same code.

**Comment by pardvm** [comment]
I'm not sure to understand when you say "_If you configure this in the config file_": Though all my servers have the same config file, I'm not configuring `processDuringTrackingRequest` in the config file, I mean that I use the UI to disable it:
![image](https://user-images.githubusercontent.com/25912392/68798800-61be3c00-0657-11ea-91ac-eb5cd2983a00.png)

Also all servers are running same matomo version 3.11... I also cannot see any reason for this effect

**Comment by tsteur** [comment]
If you had a different config file on one server where for example QueuedTracking is not marked as enabled then you could have that problem. Or if you configured above setting through the config (which is also possible). I'd just make sure all config files on all servers are identical.

**Comment by pardvm** [comment]
I'm pretty sure that all my servers have the same config file so... I really don't know what's happening.
I'll provide more information when available.
Thanks

**Comment by RentYourAdmin** [comment]
Hmm i have the same problem. I have disabled via GUI, but after the queue increases to 25 it is being triggered by the tracking request.
I have 3 webservers and 2 database server and a sentinel cluster.

**Comment by snake14** [comment]
@RentYourAdmin I'm sorry to hear that you're experiencing difficulty. The main thing is to make sure that all servers running the Matomo application have the `processDuringTrackingRequest` setting disabled, especially the servers that track requests.
